addons/portal/models/mail_message.py

```diff
@@ -2,26 +2,99 @@
+from odoo.http import request
+from odoo.tools import format_datetime
 
 
 class MailMessage(models.Model):
     _inherit = 'mail.message'
 
-    def portal_message_format(self):
+    def portal_message_format(self, options=None):
+        """ Simpler and portal-oriented version of 'message_format'.
+
+        :param dict options: options, used notably for inheritance and adding
+          specific fields or properties to compute;
+
+        :return list: list of dict, one per message in self
+        """
         self.check_access_rule('read')
-        return self._portal_message_format([
-            'id', 'body', 'date', 'author_id', 'email_from',  # base message fields
-            'message_type', 'subtype_id', 'is_internal', 'subject',  # message specific
-            'model', 'res_id', 'record_name',  # document related
-        ])
-
-    def _portal_message_format(self, fields_list):
-        vals_list = self._message_format(fields_list, legacy=True)
-        message_subtype_note_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
-        IrAttachmentSudo = self.env['ir.attachment'].sudo()
-        for vals in vals_list:
-            vals['is_message_subtype_note'] = message_subtype_note_id and (vals.get('subtype_id') or [False])[0] == message_subtype_note_id
-            for attachment in vals.get('attachment_ids', []):
-                if not attachment.get('access_token'):
-                    attachment['access_token'] = IrAttachmentSudo.browse(attachment['id']).generate_access_token()[0]
+        return self._portal_message_format(
+            self._portal_get_default_format_properties_names(options=options)
+        )
+
+    def _portal_get_default_format_properties_names(self, options=None):
+        return {
+            'attachment_ids',
+            'author_avatar_url',
+            'author_id',
+            'body',
+            'date',
+            'id',
+            'is_internal',
+            'is_message_subtype_note',
+            'published_date_str',
+            'subtype_id',
+        }
+
+    def _portal_message_format(self, properties_names):
+        """ Format messages for portal frontend.
+
+        When asking for attachments: ensure an access token is present then
+        access them (using sudo).
+
+        :param set properties_names: set of fields / properties to compute
+
+        :return list: list of dict, one per message in self
+        """
+        message_to_attachments = {}
+        if 'attachment_ids' in properties_names:
+            properties_names.remove('attachment_ids')
+            self.sudo().attachment_ids.generate_access_token()
+            related_attachments = {
+                att_read_values['id']: att_read_values
+                for att_read_values in self.env['ir.attachment'].sudo().browse(
+                    self.sudo().attachment_ids.ids
+                ).read(
+                    ["access_token", "checksum", "id", "mimetype", "name", "res_id", "res_model"]
```

Comment:
Here too `search_read` for a free query? I guess you don't need to _read_format later this way too

Reply:
It is not a search but a simple read (using a browse to create a recordset).